sklearn/linear_model/cd_fast.pyx

```diff
@@ -369,169 +600,190 @@ def sparse_enet_coordinate_descent(floating [:] w,
+                    w[ii] = 0.
                 else:
-                    w[ii] = fsign(tmp) * fmax(fabs(tmp) - alpha, 0) \
-                            / (norm_cols_X[ii] + beta)
+                    w[ii] = fsign(tmp) * fmax(fabs(tmp) - alpha, 0.) \
+                            / (norm2_cols_X[ii] + beta)
 
-                if w[ii] != 0.0:
-                    # R -=  w[ii] * X[:,ii] # Update residual
+                d_w_ii = w[ii] - w_ii
+
+                if w[ii] != w_ii:
+                    # R +=  d_w_ii * X[:,ii] # Update residual
                     for jj in range(startptr, endptr):
-                        R[X_indices[jj]] -= X_data[jj] * w[ii]
+                        R[X_indices[jj]] -= X_data[jj] * d_w_ii
 
                     if center:
-                        for jj in range(n_samples):
-                            R[jj] += X_mean_ii * w[ii]
+                        R_shift -= X_mean_ii * d_w_ii
 
                 # update the maximum absolute coefficient update
-                d_w_ii = fabs(w[ii] - w_ii)
+                d_w_ii = fabs(d_w_ii)
                 if d_w_ii > d_w_max:
                     d_w_max = d_w_ii
 
-                if w[ii] > w_max:
-                    w_max = w[ii]
-            if w_max == 0.0 or d_w_max / w_max < d_w_tol or n_iter == max_iter - 1:
-                # the biggest coordinate update of this iteration was smaller than
-                # the tolerance: check the duality gap as ultimate stopping
-                # criterion
-
-                # sparse X.T / dense R dot product
-                if center:
-                    R_sum = 0.0
-                    for jj in range(n_samples):
-                        R_sum += R[jj]
-
-                for ii in range(n_features):
-                    X_T_R[ii] = 0.0
-                    for jj in range(X_indptr[ii], X_indptr[ii + 1]):
-                        X_T_R[ii] += X_data[jj] * R[X_indices[jj]]
-
-                    if center:
-                        X_T_R[ii] -= X_mean[ii] * R_sum
-                    XtA[ii] = X_T_R[ii] - beta * w[ii]
+                if fabs(w[ii]) > w_max:
+                    w_max = fabs(w[ii])
 
-                if positive:
-                    dual_norm_XtA = max(n_features, &XtA[0])
-                else:
-                    dual_norm_XtA = abs_max(n_features, &XtA[0])
+            # Variable screening
+            if (w_max == 0.0 or
+                d_w_max / w_max < d_w_tol or
+                n_iter == max_iter - 1 or
+                screening > 0 and n_iter % screening == 0) and n_iter != 0:
+                # the biggest coordinate update of this iteration was smaller
+                # than the tolerance: check the duality gap as ultimate
+                # stopping criterion
+                gap = sparse_enet_duality_gap(n_samples, n_features,
+                                              <floating*> X_data.data,
+                                              <int*> X_indices.data,
```

Comment:
Have you considered casting these at the beginning ? I suspect this adds a python overhead, although you may have profiled this.